### DISCUSSION TITLE
Configure multiple hosts at runtime

### DISCUSSION THREAD

**Discussion by MithrilMan** [dotnet/aspnetcore#28228] [Q&A] [answered]
Hello,
I've a need regarding the creation of multiple hosts at runtime, this is my scenario:

I start with the main host using hostbuilder, where i setup services and start listening for tcp connections
based on configurations (e.g. argument or settings) I need to spawn other hosts to host different service, like for example
2 different swaggerui with different controllers, and another host to serve a blazor server.

All these hosts should share the same IServiceProvider because some components are singletons and some of the singletons are Open generics registration and as I discovered currently there is no way to create kind of child serviceproviders that share the same open generic singletons.

So I'd like to understand how can I start multiple hosts doing different things on different ports.
My requirement are because the architecture has to be thought as an extendible application (plugins).

If a single host can implement web api (controllers) on different ports using a single kestrel instance, and thus a single Host, I could adapt my code to that, so please tell me even if this last scenario is possible.

What are the alternatives?

**Comment by davidfowl** [comment, marked as answer]
One host maps to one service provider and while it may be possible to hack that, it's an assumption baked into the host (especially since ConfigureServices is a first class method on the IHostBuilder).

> If a single host can implement web api (controllers) on different ports using a single kestrel instance, and thus a single Host, I could adapt my code to that, so please tell me even if this last scenario is possible.

Yes you can listen to multiple ports and have controllers (or any endpoint really) only be valid for a particular host.

**Reply by MithrilMan** [reply]
Thanks I'm going to try it, I suppose that in order to be pluggable I'll have to create a component that store my configurations (endpoint/port, swagger name, etc..) and then use it to configure kestrel and what's needed, right before building the host.
Will ping back about it.

**Reply by davidfowl** [reply]
You can also change ports at runtime now in .NET 5 with Kestrel though there's no API to add/remove endpoints, it's currently configuration based.

**Reply by MithrilMan** [reply]
my app is configuration based too, I suppose I could anyway disable controllers at runtime just changing a flag on some Filter, the endpoint would still be open but wouldn't serve anything.
Actually I'm still on core 3.1 but planning to move to 5

**Reply by davidfowl** [reply]
Here's an example of dynamically adding a controller (same applies for removal) https://gist.github.com/davidfowl/46f470e79e23fe393e48aa4e34704357

**Reply by MithrilMan** [reply]
basically that code, instead of relying on AddControllers and custom filters/convention to discover controller types, add them directly to the application part, correct?
Seems fine thanks, even if actually the discovery in place in my app works fine, but probably I can make use of it to out-out some controllers (relying on autodiscovery to include them all but then remove the ones I don't want to be served using that dynamic application part)
Thanks for the example

P.S.
I'm using this with your bedrock framework, good job ;)
